### ISSUE TITLE
Estimating effort to support non-linear analysis (without the OS part)

### ISSUE THREAD

**Issue by rozsasarpi** [barischrooneyj/bridge-sim#166]
* Let's assume that the OS model is non-linear, e.g. cracking of concrete (material non-linearity) is considered.
* => The principle of superposition is not applicable anymore.
* => Load history matters.
* we would still like to have similar main outcomes as currently, e.g. time series of various structural responses at various points (sensors) induced by a prescribed load history.

*How large restructuring would be needed to support this (the effort without figuring out and validating the non-linear analysis parts of OS)?*

* I think it influences only the post-processing of the results (besides the small changes needed to generate the tcl files for the non-linear analysis).
* I think it would make the post-processing simpler as there would be just a single gigantic (pseudo)time-dependent non-linear analysis and time-series would be directly collected from OS.

Why I'm asking this:

- it would be great to see how the results change if we can discard the currently strongest assumption: linearity.
- our colleagues from TNO ICT might take this as an example/use case to test distributed computing and/or high-performance computing (they need a realistic case where running on a single, common computer is not sufficient, and the getting the results would have real value).

**Comment by barischrooneyj** [comment]
https://docs.google.com/document/d/1NkoJ0_epm1Q6jytnDJwCU0-gLR1LTpwoRMdW-IwARS8/edit?usp=sharing

Outline of all steps to be done, could be done in 2 days if it's planned ahead of time, there is just something that needs a bit of thought: how to structure the time series of loads in OpenSees, where nodes are placed and how to convert from traffic to loads on these nodes

**Comment by rozsasarpi** [comment]
Great, thanks a lot! It is a surprisingly low estimate, it would take more time to set up and validate the non-linear OS model. The info your provided is more than enough for now. I'll get back to this issue after a few discussions with TNO colleagues.

**Comment by barischrooneyj** [comment]
Yes I think the work is mostly on the OpenSees side. Otherwise it's primarily data transformations, which can re-use some existing functionality e.g. conversion of traffic -> loads.

**Comment by rozsasarpi** [comment]
@slobbeat  Could you take a look at [this paper](https://www.sciencedirect.com/science/article/pii/S0168874X15000074#bib4) that describes the non-linear layered shell element available in OpenSees? First, let's put aside the effort and potential computational issues with a non-linear analysis, and try to assess if this element would suit our needs. After skimming through the paper I think it would be good for us.

**Comment by slobbeat** [comment]
Looks good. Especially for the slab bridges this may work. For the girder type bridge decks, I'm not so sure yet. Here, the explicit modelling of the stirrups can be relevant (instead of smearing it out). 
The concrete material model is basically a fixed crack model, with a constant shear retention at the onset of cracking. That's certainly not the best option available. But if it is open source, we may change this later.

Questions:
- Are these element types/material models also available for quasi-static loading?
- What is the common programming language in OpenSees element/material libraries?

**Comment by rozsasarpi** [comment]
* "Are these element types/material models also available for quasi-static loading?" I guess so, when we used OS with Quanxin, we used non-linear material models under quasi-static loading.

* "What is the common programming language in OpenSees element/material libraries?"
I only know it for the general OS code base: "Written in C++, C and Fortran (C++ being the binding language)". If I remember correctly, for the element and material libraries you have to use C. You can check it [here](https://github.com/OpenSees/OpenSees).
